helix-view/src/document.rs

```diff
@@ -140,46 +140,66 @@ use helix_lsp::lsp;
 use url::Url;
 
 impl Document {
-    pub fn new(text: Rope) -> Self {
-        let changes = ChangeSet::new(&text);
-        let old_state = None;
+    pub fn new() -> Self {
+        let text = Rope::from("\n");
 
-        Self {
-            id: DocumentId::default(),
-            path: None,
-            text,
-            selections: HashMap::default(),
-            mode: Mode::Normal,
-            restore_cursor: false,
-            syntax: None,
-            language: None,
-            changes,
-            old_state,
-            diagnostics: Vec::new(),
-            version: 0,
-            history: Cell::new(History::default()),
-            last_saved_revision: 0,
-            language_server: None,
-        }
+        Self::from(text)
     }
 
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    pub fn open(path: PathBuf) -> Result<Self, Error> {
+        use std::{
+            fs::File,
+            io::{BufReader, Read},
+        };
 
-        let doc = if !path.exists() {
-            Rope::from("\n")
-        } else {
-            let file = File::open(&path).context(format!("unable to open {:?}", path))?;
-            let mut doc = Rope::from_reader(BufReader::new(file))?;
-            // add missing newline at the end of file
-            if doc.len_bytes() == 0 || doc.byte(doc.len_bytes() - 1) != b'\n' {
-                doc.insert_char(doc.len_chars(), '\n');
+        if !path.exists() {
+            return Ok(Self::new());
+        }
+
+        let mut buf = vec![0u8; 1 << 14];
+        let mut buf_str = String::with_capacity(1 << 14);
+        let mut builder = RopeBuilder::new();
+        let mut file = BufReader::new(File::open(&path)?);
+        let mut encoding = {
+            if let Ok(n) = file.read(&mut buf) {
+                let mut encoding_detector = chardetng::EncodingDetector::new();
+                encoding_detector.feed(&buf, false);
+
+                let encoding = encoding_detector.guess(None, true);
+
+                let (temp, ..) = encoding.decode(&buf);
+                buf = temp.as_bytes().to_vec();
+                buf_str.clear();
```

Comment:
There is no need for buf_str: https://doc.rust-lang.org/std/str/fn.from_utf8.html